packages/smart/contracts/turbo/IMarketFactory.sol

```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.7.6;
+pragma abicoder v2;
+
+import "../balancer/BPool.sol";
+import "../libraries/IERC20.sol";
+import "./TurboShareTokenFactory.sol";
+import "./FeePot.sol";
+
+abstract contract IMarketFactory is TurboShareTokenFactory {
+    using SafeMathUint256 for uint256;
+
+    // event MarketCreated(uint256 id, address creator, uint256 endTime, ...);
+    // event MarketResolved(uint256 id, ...);
+    event SharesMinted(uint256 id, uint256 amount, address receiver);
+    event SharesBurned(uint256 id, uint256 amount, address receiver);
+    event WinningsClaimed(uint256 id, uint256 amount, address receiver);
+
+    IERC20 public collateral;
+    FeePot public feePot;
+    uint256 public stakerFee;
+    uint256 public creatorFee;
+
+    struct Market {
+        address creator;
+        TurboShareToken[] shareTokens;
+        uint256 endTime;
+        TurboShareToken winner;
+    }
+    Market[] internal markets;
+
+    uint256 private constant MAX_UINT = 2**256 - 1;
+
+    constructor(
+        IERC20 _collateral,
+        FeePot _feePot,
+        uint256 _stakerFee,
+        uint256 _creatorFee
+    ) {
+        collateral = _collateral;
+        feePot = _feePot;
+        stakerFee = _stakerFee;
+        creatorFee = _creatorFee;
+
+        _collateral.approve(address(_feePot), MAX_UINT);
+    }
+
+    // function createMarket(address _creator, uint256 _endTime, ...) public returns (uint256);
+
+    function resolveMarket(uint256 _id) public virtual;
+
+    function getMarket(uint256 _id) public view returns (Market memory) {
+        return markets[_id];
+    }
+
+    //    function getMarketDetails(uint256 _id) public view returns (MarketDetails memory);
```

Comment:
It would be nice if this was required in concrete impls somehow but I am not sure how to do it.

Reply:
If we return a blob of bytes and add encode/decode helper methods. Then we would add that blob to getMarket. I went with this route to avoid overcomplicating the design while we're still figuring out what makes sense.